imucapture/src/main.cpp

```diff
@@ -76,245 +164,178 @@ int kbhit (){
+		throw std::runtime_error("Could not open port. Aborting.");
+	}
+	
+	std::cout << "Putting device into configuration mode..." << std::endl;
+	if (!device.gotoConfig()) {
+		throw std::runtime_error("Could not put device \
+			into configuration mode. Aborting.");
+	}
 
-			string portName;
-			cin.ignore();
-			cin >> portName;
+	// Request the device Id to check the device type
+	mtPort.setDeviceId(device.getDeviceId());
+
+	// Check if we have an MTi / MTx / MTmk4 device
+	if (!mtPort.deviceId().isMt9c() &&
+		!mtPort.deviceId().isLegacyMtig() &&
+		!mtPort.deviceId().isMtMk4() &&
+		!mtPort.deviceId().isFmt_X000())
+	{
+		throw std::runtime_error("No MTi / MTx / MTmk4 device found. \
+			Aborting.");
+	}
 
-			int baudRate;
-			cout << "Please enter baud rate (eg. 115200): ";
-			cin.ignore();
-			cin >> baudRate;
+	std::cout << "Found a device with id: "
+		<< mtPort.deviceId().toString().toStdString()
+		<< " @ port: "	 << mtPort.portName().toStdString()
+		<< ", baudrate: "  << mtPort.baudrate() << std::endl;
+
+	// Print information about detected MTi / MTx / MTmk4 device
+	std::cout << "Device: " << device.getProductCode().toStdString()
+		<< " opened." << std::endl;
+
+	// Note the differences between MTix and MTmk4
+	std::cout << "Configuring the device..." << std::endl;
+	if (mtPort.deviceId().isMt9c() || mtPort.deviceId().isLegacyMtig())
+	{
+		// output orientation data
+		XsOutputMode outputMode = XOM_Orientation;
+
+		// output orientation data as quaternion
+		XsOutputSettings outputSettings = XOS_OrientationMode_Quaternion;
+
+		// set the device configuration
+		if (!device.setDeviceMode(outputMode, outputSettings))
+			throw std::runtime_error("Could not configure MT device. \
+				Aborting.");
+			
+	}
+	else if (mtPort.deviceId().isMtMk4() || mtPort.deviceId().isFmt_X000())
+	{
+		XsOutputConfiguration quat(XDI_Quaternion, 100);
+		XsOutputConfigurationArray configArray;
+		configArray.push_back(
+			XsOutputConfiguration(XDI_LatLon | XDI_SubFormatDouble, 100)
+		);
+		configArray.push_back(
+			XsOutputConfiguration(XDI_AltitudeEllipsoid, 100)
+		);
+		configArray.push_back(
+			XsOutputConfiguration(XDI_Quaternion | XDI_CoordSysNed, 100)
+		);
+		configArray.push_back(
+			XsOutputConfiguration(XDI_VelocityXYZ | XDI_CoordSysNed, 100)
+		);
+
+		configArray.push_back(quat);
+
+		if (!device.setOutputConfiguration(configArray))
+			throw std::runtime_error("Could not configure MTmk4 device. \
+				Aborting.");
+	}
+	else
+	{
+		throw std::runtime_error("Unknown device while configuring. Aborting.");
+	}
 
-			XsPortInfo portInfo(portName, XsBaud::numericToRate(baudRate));
-			portInfoArray.push_back(portInfo);
-		}
+	std::cout << "Putting device into measurement mode..." << std::endl;
+	if (!device.gotoMeasurement()) {
+		throw std::runtime_error("Could not put device into measurement mode. \
+			Aborting.");
+	}
+}
 
-		// Use the first detected device
-		auto mtPort = portInfoArray.at(0);
+XsPortInfo scan_usb_devices()
```

Comment:
Could the return type be auto-deduced?

Reply:
Should be able to. `scan_usb_devices()` simply does what used to be there and just returns the same thing.